primap2/csg/_strategies/global_least_squares.py

```diff
@@ -0,0 +1,267 @@
+        `SubstitutionStrategy` is used.
+
+        Parameters
+        ----------
+        ts
+            Base timeseries. Missing data (NaNs) in this timeseries will be filled.
+            This function does not modify the data in ts.
+        fill_ts
+            Fill timeseries. Data from this timeseries will be used (possibly after
+            modification) to fill missing data in the base timeseries.
+            This function does not modify the data in fill_ts.
+        fill_ts_repr
+            String representation of fill_ts. Human-readable short representation of
+            the fill_ts (e.g. the source).
+
+        Returns
+        -------
+            filled_ts, descriptions. filled_ts contains the result, where missing
+            data in ts is (partly) filled using scaled data from fill_ts.
+            descriptions contains information about which years were affected and
+            filled how.
+        """
+
+        filled_mask = ts.isnull() & ~fill_ts.isnull()
+        time_filled = filled_mask["time"][filled_mask].to_numpy()
+
+        if time_filled.any():
+            # check of we have overlap. if not use substitution strategy
+            # this might not be necessary because we initialize the LS algorithm with 1,
+            # but better make it explicit
+            overlap = ts.notnull() & fill_ts.notnull()
+            if overlap.any():
+                if self.allow_shift:
+                    #  pass on to lstsq based strategy
+                    strategy = primap2.csg.GlobalLSlstsqStrategy()
+                    filled_ts, descriptions = strategy.fill(
+                        ts=ts,
+                        fill_ts=fill_ts,
+                        fill_ts_repr=fill_ts_repr,
+                    )
+                else:
+                    e = fill_ts[overlap.data].data
+                    e_ref = ts[overlap.data].data
+                    a0 = [1]  #  start with 1 as scaling factor
+                    res = least_squares(
+                        self.factor_mult, a0, jac=self.jac, args=(e, e_ref)
+                    )
+
+                    fill_ts_h = fill_ts * res["x"][0]
+                    filled_ts = xr.core.ops.fillna(ts, fill_ts_h, join="exact")
+
+                    descriptions = [
+                        primap2.ProcessingStepDescription(
+                            time=time_filled,
+                            description="filled with least squares matched data from "
+                            f" {fill_ts_repr}. Factor={res['x'][0]:0.3f}",
+                            function=self.type,
+                            source=fill_ts_repr,
+                        )
+                    ]
+            else:
+                strategy = primap2.csg.SubstitutionStrategy()
+                filled_ts, descriptions = strategy.fill(
+                    ts=ts,
+                    fill_ts=fill_ts,
+                    fill_ts_repr=fill_ts_repr,
+                )
```

Comment:
```suggestion
            else:
                raise StrategyUnableToProcess(reason="No overlap between timeseries, can't match")
```

I think the choice what to do if Ls matching is not feasible should be left to the user.

Reply:
Would the user have to add another item to the strategy list? As we would have to do that for every source / scenario combination in PRIMAP-hist individually this would make the strategy list very long with repetitive information. Or do we have a different way to specify this?
We could also add a fallback parameter to the strategy

Reply:
You can simply have a fallback

```python
strategy_definition = StrategyDefinition([
    # For EDGAR and ANDREWS, shifting doesn't work well 
    ({"source": ["EDGAR", "ANDREWS"]}, GlobalLSStrategy(allow_shift=False)),
    # All other sources work well with shifting
    ({"source": ["CRF2021", "CRF2022", "CRF2023"]}, GlobalLSStrategy(allow_shift=True)),
    # If the least-squares strategy fails, use simple substitution instead
    ({}, SubstitutionStrategy())
])
```
Of course, using the default selector like this means if you forget to include a strategy in the sources list for both GlobalLSStrategies, it will just use `SubstitutionStrategy` instead of raising an error. To guard against this and maybe be a bit more explicit you could write it like this:
```python
primap_hist_sources = ["EDGAR", "ANDREWS", "CRF2021", "CRF2022", "CRF2023"]
no_shift_sources = ["EDGAR", "ANDREWS"]

strategy_definition = StrategyDefinition([
    # For EDGAR and ANDREWS, shifting doesn't work well 
    ({"source": no_shift_sources}, GlobalLSStrategy(allow_shift=False)),
    # All other sources work well with shifting
    ({"source": [source for source in primap_hist_sources if source not in no_shift_sources]}, GlobalLSStrategy(allow_shift=True)),
    # If the least-squares strategy fails, use simple substitution instead
    ({"source": primap_hist_sources}, SubstitutionStrategy())
])
```
Or are the rules much more complex than I imagine?

Reply:
I'll remove all internal fallbacks (to be consistent also the fallback which is triggered when negative data are created) and will deal with them through configuration. It might make debugging harder because a fallback might also be a default, but with a more explicit configuration I think we can deal with that.